FreeRTOS/Test/Target/tests/smp/fr09_interruptWaitCritical/test.c

```diff
@@ -0,0 +1,210 @@
+ *     yield to taskC which has equal priority.
+ */
+
+/* Kernel includes. */
+#include "FreeRTOS.h" /* Must come first. */
+#include "task.h"     /* RTOS task related API prototypes. */
+
+#include <string.h>
+
+#include "bsl.h"
+#include "unity.h" /* unit testing support functions */
+/*-----------------------------------------------------------*/
+
+/* Priorities at which the tasks are created.  The max priority can be specified
+ * as ( configMAX_PRIORITIES - 1 ). */
+#define mainTASK_A_PRIORITY    ( tskIDLE_PRIORITY + 3 )
+#define mainTASK_B_PRIORITY    ( tskIDLE_PRIORITY + 2 )
+#define mainTASK_C_PRIORITY    ( tskIDLE_PRIORITY + 1 )
+
+static void vPrvTaskA( void * pvParameters );
+static void vPrvTaskB( void * pvParameters );
+static void vPrvTaskC( void * pvParameters );
+/*-----------------------------------------------------------*/
+
+#if configNUMBER_OF_CORES != 2
+    #error Require two cores be configured for FreeRTOS
+#endif /* if configNUMBER_OF_CORES != 2 */
+
+#if traceTASK_SWITCHED_IN != test_fr9TASK_SWITCHED_IN
+    #error Need to include testConfig.h in FreeRTOSConfig.h
+#endif /* if traceTASK_SWITCHED_IN != test_fr2TASK_SWITCHED_IN */
+/*-----------------------------------------------------------*/
+
+static TaskHandle_t xTaskAHandler;
+static TaskHandle_t xTaskBHandler;
+static TaskHandle_t xTaskCHandler;
+static BaseType_t xTestFailed = pdFALSE;
+static BaseType_t xAllTasksHaveRun = pdFALSE;
+static BaseType_t xTaskAHasEnteredCriticalSection = pdFALSE;
+static BaseType_t xTaskAHasExitedCriticalSection = pdFALSE;
+static BaseType_t xTaskBHasEnteredCriticalSection = pdFALSE;
+
+static uint32_t ulTaskSwitchCount = 0;
+static BaseType_t xTaskARan = pdFALSE;
+static BaseType_t xTaskBRan = pdFALSE;
+static BaseType_t xTaskCRan = pdFALSE;
+/*-----------------------------------------------------------*/
+
+void test_fr9TASK_SWITCHED_IN( void )
+{
+    UBaseType_t uxIdx, uxNumTasksRunning;
+    TaskStatus_t taskStatus[ 16 ];
+    UBaseType_t uxTaskStatusArraySize = 16;
+    unsigned long ulTotalRunTime;
+
+    if( ( ( xAllTasksHaveRun == pdFALSE ) && ( xTestFailed == pdFALSE ) ) )
+    {
+        uxNumTasksRunning = uxTaskGetSystemState( ( TaskStatus_t * const ) &taskStatus, uxTaskStatusArraySize, &ulTotalRunTime );
+
+        for( uxIdx = 0; uxIdx < uxNumTasksRunning; uxIdx++ )
+        {
+            if( ( strcmp( taskStatus[ uxIdx ].pcTaskName, "TaskA" ) == 0 ) && ( taskStatus[ uxIdx ].eCurrentState == eRunning ) )
+            {
+                xTaskARan = pdTRUE;
+            }
+
+            if( ( strcmp( taskStatus[ uxIdx ].pcTaskName, "TaskB" ) == 0 ) && ( taskStatus[ uxIdx ].eCurrentState == eRunning ) )
+            {
+                xTaskBRan = pdTRUE;
+            }
+
+            if( ( strcmp( taskStatus[ uxIdx ].pcTaskName, "TaskC" ) == 0 ) && ( taskStatus[ uxIdx ].eCurrentState == eRunning ) )
+            {
+                xTaskCRan = pdTRUE;
+            }
+        }
+
```

Comment:
We should add two checkers here:
1. When task C is running, task A&B should be run before.
2. Task C is running with task A only before task A ends its job.